File: web/tuning/battery.js
'use strict';
window.batteryDashboard = {
  groups: {power:'01 · 배터리 전력 / 좌우 출력 요구',voltage:'02 · 팩 전압',current:'03 · 팩 전류',delta:'04 · 토크벡터링 ΔP',yaw:'05 · 목표 / 실제 yaw',throttle:'06 · 스로틀',speed:'07 · 차속',cells:'08 · 최고 / 최저 셀 전압',imbalance:'09 · 셀 전압 편차',temperature:'10 · 온도',soc:'11 · 잔량'},
  render(values,status,view,selected,showingCurrent) {
    const unit=(v,u,d=1)=>finite(v)?format(v,d)+' '+u:'미수신';
    const on=v=>v===true?'ON':v===false?'OFF':'미수신';
    $('bmsContext').textContent=showingCurrent?'현재 수신값':`선택 시점 · ${selected?time(selected.time_ms):'기록 없음'}`;
    $('packWatts').textContent=unit(values.battery_power_w,'W',0);
    $('packKw').textContent=unit(values.battery_power_kw,'kW',3);
    $('packVoltage').textContent=format(values.battery_pack_voltage_v,1);
    $('packCurrent').textContent=format(values.battery_current_a,1);
    $('packSoc').textContent=format(values.battery_soc_pct,1);
    $('commandTotal').textContent=unit(values.power_command_total_kw,'kW',3);
    $('commandSides').textContent=`좌 ${unit(values.power_left_kw,'kW',3)} / 우 ${unit(values.power_right_kw,'kW',3)} · 요구값`;
    $('batterySpeed').textContent=unit(values.speed_kmh,'km/h',2);
    let hasChange=false;
    for(const [key,id,u,d] of [['battery_pack_voltage_v','voltageChange','V',1],['battery_current_a','currentChange','A',1],['battery_power_w','powerChange','W',0]]){
      const first=view.rows.find(r=>r.id&&finite(r.values[key])&&(!selected||r.time_ms<=selected.time_ms));
      const valid=finite(values[key])&&first;
      const change=valid?values[key]-first.values[key]:null;
      $(id).textContent=valid?`${change>0?'+':''}${format(change,d)} ${u}`:'—';
      hasChange=hasChange||Boolean(valid);
    }
    $('changeContext').textContent=hasChange?'표시 구간 기준 · 포인터로 비교 시점 선택':'구간 내 유효값 대기';
    $('bmsState').textContent=status.bms_online?(status.bms_state||'수신 중'):'BMS 미수신';
    $('bmsState').classList.toggle('alarm',status.bms_fault===true);
    $('mos').textContent=`충전 ${on(status.bms_charge_mos_on)} / 방전 ${on(status.bms_discharge_mos_on)}`;
    $('load').textContent=`충전기 ${on(status.bms_charger_present)} / 부하 ${on(status.bms_load_present)}`;
    $('temps').textContent=`${unit(values.bms_temp_min_c,'°C')} / ${unit(values.bms_temp_max_c,'°C')}`;
    $('extra').textContent=`${on(status.bms_balancing)} / ${unit(status.bms_remaining_capacity_ah,'Ah',2)}`;
    $('cycles').textContent=format(status.bms_cycle_count,0);
    $('alarm').textContent=status.bms_online?(status.bms_fault===true?`경고 · ${status.bms_alarm_summary||status.bms_alarm_hex||'BMS 알람'}`:status.bms_fault===false?'보고된 알람 없음':'알람 상태 미수신'):'미수신';
    $('alarm').classList.toggle('alarm',status.bms_fault===true);
    $('bmsAge').textContent=finite(status.bms_age_ms)?`최근 BMS 프레임 ${format(status.bms_age_ms,0)} ms 전 · 전압/전류 개별 시각은 미제공`:'CAN 수신 대기';
    $('bmsSource').textContent=[status.bms_model,status.bms_protocol].filter(Boolean).join(' · ')||'BMS 수신 대기';
    const cells=status.bms_cell_voltages_v||[],validCells=cells.filter(finite);
    $('cellCount').textContent=status.bms_online?`${validCells.length}개 수신 / ${finite(status.bms_cell_count)?status.bms_cell_count:'?'}셀`:'미수신';
    $('cellSummary').textContent=`최저 ${unit(values.bms_min_cell_voltage_v,'V',3)} / 최고 ${unit(values.bms_max_cell_voltage_v,'V',3)} · 편차 ${unit(values.bms_cell_delta_mv,'mV',0)}`;
    $('cells').replaceChildren();
    const min=Math.min(...validCells),max=Math.max(...validCells);
    cells.forEach((v,i)=>{const cell=document.createElement('div');cell.className='cell';cell.classList.toggle('low',finite(v)&&v===min);cell.classList.toggle('high',finite(v)&&v===max&&min!==max);cell.append(`셀 ${i+1}`);const value=document.createElement('b');value.textContent=unit(v,'V',3);cell.append(value);$('cells').append(cell);});
  }
};

File: gateway/battery.py
"""Measured BMS channels and clearly labelled TV commands for battery analysis."""
import math
from tuning import project as tuning_project

# key, label, unit, group, validity flags (same channel contract as /tuning).
CHANNELS = (
    ('battery_power_kw', '배터리 전체 전력 · V×I', 'kW', 'power', ('bms_online',)),
    ('power_command_total_kw', '좌우 출력 요구 합계 · 실측 아님', 'kW', 'power', ('tqv_internal_online',)),
    ('power_left_kw', '좌측 출력 요구', 'kW', 'power', ('tqv_internal_online',)),
    ('power_right_kw', '우측 출력 요구', 'kW', 'power', ('tqv_internal_online',)),
    ('battery_pack_voltage_v', '팩 전압', 'V', 'voltage', ('bms_online',)),
    ('battery_current_a', '팩 전류 · BMS 부호 유지', 'A', 'current', ('bms_online',)),
    ('battery_soc_pct', '배터리 잔량', '%', 'soc', ('bms_online',)),
    ('bms_max_cell_voltage_v', '최고 셀 전압', 'V', 'cells', ('bms_online',)),
    ('bms_min_cell_voltage_v', '최저 셀 전압', 'V', 'cells', ('bms_online',)),
    ('bms_cell_delta_mv', '최고−최저 셀 전압 차이', 'mV', 'imbalance', ('bms_online',)),
    ('bms_temp_max_c', '최고 온도', '°C', 'temperature', ('bms_online',)),
    ('bms_temp_min_c', '최저 온도', '°C', 'temperature', ('bms_online',)),
    ('delta_power_kw', '좌우 차동 출력 요구 ΔP', 'kW', 'delta', ('tqv_internal_online',)),
    ('desired_yaw_rad_s', '목표 yaw rate', 'rad/s', 'yaw', ('tqv_internal_online',)),
    ('yaw_rate_rad_s', '현재 yaw rate', 'rad/s', 'yaw', ('imu_online', 'imu_ok')),
    ('yaw_error_rad_s', 'STM yaw 오차', 'rad/s', 'yaw', ('tqv_internal_online',)),
    ('tps_pct', '스로틀 개도량', '%', 'throttle', ('tps_online', 'tps_ok')),
    ('speed_kmh', '좌우 평균 RPM 환산 차속', 'km/h', 'speed', ('speed_online',)),
    ('battery_power_w', '배터리 전체 전력 · W', 'W', 'readouts', ('bms_online',)),
)
INPUTS = {c[0]: (c[0],) for c in CHANNELS if c[0].startswith(('battery_', 'bms_'))}
INPUTS.update({
    'battery_pack_voltage_v': ('battery_pack_voltage_v', 'pack_voltage_v'),
    'battery_current_a': ('battery_current_a', 'pack_current_a'),
    'battery_soc_pct': ('battery_soc_pct', 'soc_pct', 'battery_pct', 'battery_soc'),
})
STATUS_KEYS = ('bms_fault', 'bms_charge_mos_on', 'bms_discharge_mos_on',
               'bms_balancing', 'bms_charger_present', 'bms_load_present',
               'bms_cell_count', 'bms_cycle_count', 'bms_remaining_capacity_ah',
               'bms_state', 'bms_alarm_hex', 'bms_alarm_summary', 'bms_model',
               'bms_protocol', 'bms_cell_voltages_v')


def finite(value):
    return isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value)


def with_power(data):
    """Use V and I from the same received packet, never mixed-age held values."""
    result = dict(data)
    voltage = data.get('battery_pack_voltage_v', data.get('pack_voltage_v'))
    current = data.get('battery_current_a', data.get('pack_current_a'))
    if finite(voltage) and voltage > 0 and finite(current) and finite(voltage * current):
        result['battery_power_w'] = voltage * current
        result['battery_power_kw'] = voltage * current / 1000.0
    elif any(key in data for key in ('battery_pack_voltage_v', 'pack_voltage_v',
                                     'battery_current_a', 'pack_current_a')):
        # A partial/invalid electrical update cannot refresh an old power value.
        result['battery_power_w'] = None
        result['battery_power_kw'] = None
    else:
        power = data.get('battery_power_kw', data.get('pack_power_kw'))
        if finite(power):
            result['battery_power_kw'] = power
            result['battery_power_w'] = power * 1000.0
    return result


def project(packet):
    values = tuning_project(packet)
    source = packet.get('_tuning_values', {}) if packet.get('_battery_fields_tracked') else with_power(packet)
    for key in INPUTS:
        value = source.get(key)
        valid = bool(packet.get('bms_online')) and finite(value)
        timing = packet.get('bms_power_timing')
        if key.startswith('battery_') and isinstance(timing, list):
            valid = valid and len(timing) == 8 and finite(timing[4]) and 0 <= timing[4] < 2000
        if key in ('battery_pack_voltage_v', 'bms_max_cell_voltage_v', 'bms_min_cell_voltage_v'):
            valid = valid and value > 0
        if key == 'battery_soc_pct':
            valid = valid and 0 <= value <= 100
        if key == 'bms_cell_delta_mv':
            valid = valid and value >= 0
        values[key] = value if valid else None
    left, right = values.get('power_left_kw'), values.get('power_right_kw')
    values['power_command_total_kw'] = left + right if finite(left) and finite(right) else None
    return {key: values.get(key) for key, *_ in CHANNELS}


def status(packet):
    """Allowlisted metadata, with unavailable states represented as null."""
    online = bool(packet.get('bms_online'))
    source = packet.get('_battery_status', packet)
    result = {'bms_online': online}
    for key in ('bms_fault', 'bms_charge_mos_on', 'bms_discharge_mos_on',
                'bms_balancing', 'bms_charger_present', 'bms_load_present'):
        result[key] = bool(source[key]) if online and isinstance(source.get(key), (bool, int)) else None
    for key in ('bms_age_ms', 'bms_cell_count', 'bms_cycle_count', 'bms_remaining_capacity_ah'):
        value = packet.get(key) if key == 'bms_age_ms' else source.get(key)
        result[key] = value if online and finite(value) else None
    for key in ('bms_state', 'bms_alarm_hex', 'bms_alarm_summary', 'bms_model', 'bms_protocol'):
        result[key] = str(source.get(key) or '')[:95] if online else ''
    cells = source.get('bms_cell_voltages_v', [])
    result['bms_cell_voltages_v'] = [v if finite(v) and v > 0 else None for v in cells[:48]] if online and isinstance(cells, list) else []
    return result
